File: python/euler/bin/p0082.py
# project euler: problem 82

from collections.abc import Callable
from typing import IO


def compute(fn: Callable[..., int], fh: IO) -> str:
    def parse_data(fh: IO) -> list[list[int]]:
        # transpose matrix
        return [
            list(x)
            for x in zip(
                *[list(map(int, line.split(','))) for line in fh.read().splitlines()]
            )
        ]

    matrix = parse_data(fh)
    work = matrix[0]
    for crnt in matrix[1:]:
        work[0] += crnt[0]
        for i in range(1, len(crnt)):
            work[i] = crnt[i] + fn(work[i], work[i - 1])
        for i in reversed(range(0, len(crnt) - 1)):
            work[i] = fn(work[i], work[i + 1] + crnt[i])

    return str(sorted(work)[0])


def solve() -> str:
    from euler.lib.resource import asset_file

    fh = asset_file('p082_matrix.txt')
    result = compute(min, fh)
    fh.close()
    return result

File: typescript/euler/bin/p0082.ts
// project euler: problem 82

import { assetData } from "../lib/asset.ts";
import { range } from "../lib/util.ts";

export const compute = (
  fn: (...valus: number[]) => number,
  data: string,
): string => {
  function parseData(data: string): number[][] {
    function splitLines(str: string): string[] {
      const result = str.split(/\r?\n/);
      if (result.at(-1) === "") {
        return result.slice(0, -1);
      } else {
        return result;
      }
    }

    return splitLines(data).map((x) => x.split(",").map((y) => Number(y)));
  }

  function transpose(matrix: number[][]): number[][] {
    // Add a check process for square matrix if necessary
    return matrix[0].map((_, idx) => matrix.map((row) => row[idx]));
  }

  const matrix = transpose(parseData(data));
  const work = matrix[0];
  for (const crnt of matrix.slice(1)) {
    work[0] += crnt[0];
    for (const i of range(1, crnt.length)) {
      work[i] = crnt[i] + fn(work[i], work[i - 1]);
    }
    for (const i of range(0, crnt.length - 1).reverse()) {
      work[i] = fn(work[i], work[i + 1] + crnt[i]);
    }
  }

  return String(work.sort()[0]);
};

export const solve = (): string => {
  const data = new TextDecoder().decode(assetData("p082_matrix.txt"));
  return compute(Math.min, data);
};
